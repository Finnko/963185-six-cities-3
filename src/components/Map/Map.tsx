import L from 'leaflet';
import React from 'react';
import {MapContainer, TileLayer, Marker, Popup, useMap} from 'react-leaflet';
import {useSelector} from 'react-redux';

import {RootState} from '../../app/rootReducer';
import {selectCurrentCity} from '../../features/offers/selectors';
import {IOffer} from '../../interfaces/offer';

const LeafletProperties = {
  TILE_LAYER: 'http://{s}.tiles.maps.sputnik.ru/{z}/{x}/{y}.png',
  ATTRIBUTION: '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
};

const PinSize = {
  WIDTH: 27,
  HEIGHT: 39,
};

interface MapProps {
  offers: IOffer[];
}

interface ChangeViewProps {
  latitude: number;
  longitude: number;
  zoom: number;
}

const defaultIcon = new L.Icon({
  iconUrl: '../img/pin.svg',
  iconSize: [PinSize.WIDTH, PinSize.HEIGHT],
  iconAnchor: [PinSize.WIDTH / 2, PinSize.HEIGHT],
  popupAnchor: [0, -PinSize.HEIGHT / 3],
});

const activeIcon = new L.Icon({
  iconUrl: '../img/pin-active.svg',
  iconSize: [PinSize.WIDTH, PinSize.HEIGHT],
  iconAnchor: [PinSize.WIDTH / 2, PinSize.HEIGHT],
  popupAnchor: [0, -PinSize.HEIGHT / 3],
});

const renderPins = (offers: IOffer[], activeOffer: number | null) => offers.map(offer => {
  const {
    location: {
      latitude,
      longitude,
    },
    id,
    previewImage,
    price,
    title,
  } = offer;

  return (
    <Marker
      key={id}
      position={[latitude, longitude]}
      icon={activeOffer === id ? activeIcon : defaultIcon}
    >
      <Popup>
        <div className="place-card__image-wrapper cities__image-wrapper">
          <a href="#">
            <img className="place-card__image"
                 src={previewImage}
                 width="260"
                 height="200"
                 alt={title}
            />
          </a>
        </div>
        <h2 className="place-card__name">
          {title}
        </h2>
        <div className="place-card__price">
          <b className="place-card__price-value">â‚¬{price}</b>
          <span className="place-card__price-text">/&nbsp;night</span>
        </div>
      </Popup>
    </Marker>
  );
});

const ChangeView: React.FC<ChangeViewProps> = ({latitude, longitude, zoom}) => {
  const map = useMap();

  map.flyTo([latitude, longitude], zoom, {
    duration: 1.5, // sec
  });

  return null;
};


const Map: React.FC<MapProps> = ({offers}) => {
  const city = useSelector(selectCurrentCity);
  console.log(city);
  const {activeOffer} = useSelector((state: RootState) => state.ui);

  const {
    location: {
      latitude,
      longitude,
      zoom,
    }
  } = city;

  return (
    <div className="cities__right-section">
      <MapContainer
        center={[latitude, longitude]}
        zoom={zoom}
        className="cities__map map"
      >
        <ChangeView
          latitude={latitude}
          longitude={longitude}
          zoom={zoom}
        />
        <TileLayer
          attribution={LeafletProperties.ATTRIBUTION}
          url={LeafletProperties.TILE_LAYER}
        />
        {renderPins(offers, activeOffer)}
      </MapContainer>
    </div>
  );
};
export default Map;
