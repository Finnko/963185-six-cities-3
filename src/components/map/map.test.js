import React from 'react';
import renderer from 'react-test-renderer';
import Map from './map.jsx';

const mocks = [
  {
    coords: [52.38533, 4.89],
  }
];

it(`<Map /> should be render correctly`, () => {
  const map = renderer.create(<Map offers={mocks} isNearByView={false}/>, {
    createNodeMock: () => {
      return document.createElement(`div`);
    }
  }).toJSON();

  expect(map).toMatchSnapshot();
});
