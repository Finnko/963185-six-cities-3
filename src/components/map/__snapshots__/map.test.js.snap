// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> should be render correctly 1`] = `
<section
  className="map cities__map"
>
  <div
    id="map"
    style={
      Object {
        "height": "100%",
      }
    }
  />
</section>
`;
